### PR TITLE
Webhook Service Initial Implementation

### LINKED ISSUE #4341: Jenkins Proxy proper external validation
Jenkins Proxy is not validating properly the external requests from Github before passing them to Jenkins, let's investigate how we can make this better,

### PR DESCRIPTION
**Pull request by khrm**
Initial Implementation of Webhook service needed to resolve https://github.com/openshiftio/openshift.io/issues/4341
It uses IP to verify webhooks are coming from GH,

### CODE REVIEW

File: verification/verification.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+// New returns a verification service instance
+func New(gs *goa.Service, duration time.Duration) (Service, error) {
+	s := &service{
+		Service: gs,
+		ticker:  time.NewTicker(duration),
+	}
+	if err := s.setHookIPs(); err != nil {
+		return nil, err
+	}
+	go s.monitorHookIPs()
```
Comment (khrm) [inline verification/verification.go]: Running a monitor in Background. Better way to do this?

File: configuration/configuration.go [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+	"time"
+
+	errs "github.com/pkg/errors"
+	"github.com/spf13/viper"
+)
+
+const (
+	// Constants for viper variable names. Will be used to set
+	// default values as well as to get each value
+	varCleanTestDataEnabled = "clean.test.data"
```
Comment (sthaha) [inline configuration/configuration.go:16]: why is there a var prefix? is that intentional?

File: configuration/configuration.go [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+	return c.v.GetString(varMetricsHTTPAddress)
+}
+
+// GetDiagnoseHTTPAddress returns the address of where to start the gops handler.
+// By default GetDiagnoseHTTPAddress is 127.0.0.1:0 in devMode, but turned off in prod mode
+// unless explicitly configured
+func (c *Config) GetDiagnoseHTTPAddress() string {
+	if c.v.IsSet(varDiagnoseHTTPAddress) {
+		return c.v.GetString(varDiagnoseHTTPAddress)
+	} else if c.DeveloperModeEnabled() {
```
Comment (sthaha) [inline configuration/configuration.go:171]: `else` isn't needed

File: design/webhook.go [hunk 4]
```diff
@@ -0,0 +1,22 @@
+package design
+
+import (
+	d "github.com/goadesign/goa/design"
```
Comment (sthaha) [inline design/webhook.go:4]: isn't the convention to `. import` here?

File: verification/verification.go (hunk 1 above)
Comment (sthaha) [reply verification/verification.go]: when does this stop?

File: verification/verification.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+	for range s.ticker.C {
+		if err := s.setHookIPs(); err != nil {
+			s.Service.LogError("Error while monitoring Hook IPS")
+		}
+	}
+}
+
+func (s *service) isGithubIP(i string) bool {
+	ip := net.ParseIP(i)
+	s.lock.RLock()
```
Comment (sthaha) [inline verification/verification.go:109]: isn't it better to `defer s.lock.RUnlock()` ?

File: main.go [hunk 6]
```diff
@@ -99,6 +100,16 @@ func main() {
 	statusCtrl := controller.NewStatusController(service)
 	app.MountStatusController(service, statusCtrl)
 
+	verificationSvc, err := verification.New(service, config.GetMonitorIPDuration())
+
+	if err != nil {
+		log.Logger().Fatal("Verification Service Initialisation Failed", err)
```
Comment (sthaha) [inline main.go:114]: how does the log look like? what field are getting added to the log?

File: main.go [hunk 7]
```diff
@@ -99,6 +100,16 @@ func main() {
...
+	}
+
+	// Mount "webhook" controller
+	webhookCtrl := controller.NewWebhookController(service,
+		config, verificationSvc)
+	app.MountWebhookController(service, webhookCtrl)
 	log.Logger().Infoln("Git Commit SHA: ", app.Commit)
```
Comment (sthaha) [inline main.go:121]: may want to store the log.Logger() `logger := log.Logger().WithField("component", "main")` etc .. and reuse the `logger` object.

File: verification/verification.go (hunk 1 above)
Comment (sthaha) [reply verification/verification.go]: https://dave.cheney.net/2016/12/22/never-start-a-goroutine-without-knowing-how-it-will-stop

Comment (khrm) [reply verification/verification.go]: I have read it before but wanted a pattern where another goroutine is running in background to check for ip changes. It never stops. It uses a for a loop on the ticker.  You can send s.Ticker.Stop to stop it.

File: design/webhook.go (hunk 4 above)
Comment (khrm) [reply design/webhook.go:4]: Yes but fabric starter uses this for status service already, so I am using this.

File: configuration/configuration.go (hunk 3 above)
Comment (khrm) [reply configuration/configuration.go:171]: This is else if. Ignore this file. I need to change it to use fabric-common configuration.

File: configuration/configuration.go (hunk 2 above)
Comment (khrm) [reply configuration/configuration.go:16]: Yes. Fabric-common configuration seems to use it. No idea why and other projects using fabric-starter-goa are also using it.

File: verification/verification.go (hunk 1 above)
Comment (sthaha) [reply verification/verification.go]: why not fetch the list initally and then refetch the list only if the cache seems to not find the ip in the list? that way we dont need a go routine polling for ever?

Comment (khrm) [reply verification/verification.go]: Yes, that's a better pattern.

File: main.go (hunk 7 above)
Comment (khrm) [reply main.go:121]: This is generated code from f8-starter

File: design/webhook.go (hunk 4 above)
Comment (chmouel) [reply design/webhook.go:4]: it's. a convention for all goadesign based projects,

File: controller/webhook.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+// Forward runs the forward action.
+func (c *WebhookController) Forward(ctx *app.ForwardWebhookContext) error {
+	// WebhookController_Forward: start_implement
+
+	// Put your logic here
+
+	// WebhookController_Forward: end_implement
+	//	authURL, _ := url.Parse(c.config.GetAuthServiceURL())
+
+	u, _ := url.Parse("http://localhost:9091")
```
Comment (chmouel) [inline controller/webhook.go]: this should be configurable,

File: controller/webhook.go [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+	vs verification.Service) *WebhookController {
+	return &WebhookController{
+		Controller: service.NewController("WebhookController"),
+		config:     config}
+}
+
+// Forward runs the forward action.
+func (c *WebhookController) Forward(ctx *app.ForwardWebhookContext) error {
+	// WebhookController_Forward: start_implement
+
```
Comment (chmouel) [inline controller/webhook.go:55]: did you copy and paste from somewhere? can you remove the ocmments

File: controller/webhook.go (hunk 8 above)
Comment (chmouel) [reply controller/webhook.go]: and err handled,

File: verification/verification.go [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+// Verify verifies whether request came
+// from approved source
+func (s *service) Verify(req *http.Request) bool {
+	ip := strings.Split(req.Header.Get("X-Forwarded-For"), ",")
+	if len(ip[0]) == 0 {
+		ip = strings.Split(req.RemoteAddr, ":")
+	}
+	s.Service.LogInfo("IP", "ip", ip)
```
Comment (chmouel) [inline verification/verification.go]: can we have some more meaningful messages?

File: controller/webhook.go (hunk 9 above)
Comment (khrm) [reply controller/webhook.go:55]: No. That was generated comment from f8-starter. Removing it.

File: controller/webhook.go (hunk 8 above)
Comment (khrm) [reply controller/webhook.go]: This is configurable now. Adding err check.

File: main.go (hunk 6 above)
Comment (khrm) [reply main.go:114]: Like this:
```
{"err":"unexpected end of JSON input","fields.msg":{"errors":[{"code":"unknown_error","detail":"unexpected end of JSON input","status":"500","title":"Unknown error"}]},"file":"/home/kbaig/go/src/github.com/fabric8-services/fabric8-webhook/app/error_handler.go","func":"github.com/fabric8-services/fabric8-webhook/app.ErrorHandler.func1.1","level":"error","line":73,"msg":"uncaught error detected in ErrorHandler","pid":11254,"pkg":"github.com/fabric8-services/fabric8-webhook/app.ErrorHandler.func1","req_headers":{"Accept":["*/*"],"Accept-Encoding":["gzip, deflate"],"Accept-Language":["en-US,en;q=0.5"],"Connection":["keep-alive"],"Content-Length":["0"],"Content-Type":["text/plain;charset=UTF-8"],"Dnt":["1"],"User-Agent":["Mozilla/5.0 (X11; Fedora; Linux x86_64; rv:64.0) Gecko/20100101 Firefox/64.0"],"X-Forwarded-For":["192.30.252.1"]},"req_id":"shnqvyr8gy-4","time":"2019-01-28 16:16:44"}
```

**Review by nikhil-thomas — COMMENTED** [review]
Looks good to me.

just one comment
I see that you wrote a wrapper 'config' to wrap viper

I think in Go it might be better if can define a 'config interface'
and have a 'viper implementation' that way it will be more pluggable

In addition, we can invert the direction of dependency. Instead of our core logic being dependent on 'config wrapper', the config implementations (eg: viper) will be dependent on an interface we defined upfront.

```golang
type config interface {
    GetLogLevel() string
    GetPostgresHost() string
    GetPostgresPort() int64
}

type ViperConfig struct {
    //embeded viper or make it a field
    viper.Viper
}

func (vc *ViperConfig) GetLogLevel() string {
	return vc.GetString(varLogLevel)
}

func (c *Config) GetPostgresHost() string {
	return vc.GetString(varPostgresHost)
}

func (c *Config) GetPostgresPort() int64 {
	return vc.GetInt64(varPostgresPort)
}
```
* just my opinion, please disregard if there are things i am overlooking

**Comment by pradeepitm12** [comment]
Looks good to me.

**Pull request merged**